internal/validators/namespace.go

```diff
@@ -106,20 +106,49 @@ func (v *Namespace) handle(req *nsRequest) admission.Response {
 	return allow("")
 }
 
-func (v *Namespace) illegalExcludedNamespaceLabel(req *nsRequest) admission.Response {
-	for l := range req.ns.Labels {
-		if l == api.LabelExcludedNamespace && !config.ExcludedNamespaces[req.ns.Name] {
-			// Note: this only blocks the request if it has a newly added illegal
-			// excluded-namespace label because existing illegal excluded-namespace
-			// label should have already been removed by our reconciler. For example,
-			// even when the VWHConfiguration is removed, adding the label to a non-
-			// excluded namespace would pass but the label is immediately removed; when
-			// the VWHConfiguration is there but the reconcilers are down, any request
-			// gets denied anyway.
-			msg := fmt.Sprintf("You cannot exclude this namespace using the %q label. See https://github.com/kubernetes-sigs/hierarchical-namespaces/blob/master/docs/user-guide/concepts.md#excluded-namespace-label for detail.", api.LabelExcludedNamespace)
+// illegalIncludedNamespaceLabel checks if there's any illegal use of the
+// included-namespace label on namespaces.
+func (v *Namespace) illegalIncludedNamespaceLabel(req *nsRequest) admission.Response {
+	labelValue, hasLabel := req.ns.Labels[api.LabelIncludedNamespace]
```

Comment:
Maybe file this as a TODO and a future issue, but I think I'd also like to see this validator allow any request that doesn't _change_ these values, even if they're wrong. E.g. if this is an included namespace but the label is already missing, don't reject the request just because it continues to be missing. Just let the reconciler fix it up.

If you do want to fix this up, you'll have to add `nsOld` to `nsRequest` and then you'll say something like:

```
// Early exit if nothing has changed
if req.nsOld != nil {
  oldVal, oldHas := rq.oldNS.Labels[api.LabelIncludedNamespace]
  if oldVal == labelValue && oldHas == hasLabel {
    return allow("")
  }
}
```

Reply:
That's interesting. I agree with you that this is will happen in practice if the reconciler is working properly. However, that's a pretty strong coupling - to understand why _this_ code works correctly, you need to fully understand how a _different_ piece of code works. If that other code changes (because of a bug or because of an intentional change), this code could break. Plus it's just harder to understand.

On the other hand, if you early-exit and `return true` if these values are unchanged, now this code works _in isolation_. You don't have to care at all about what the reconciler's doing - the reconciler is now basically a fully redundant backup to this code, and as long as either of them are working as intended, HNC will continue to work.

Does that make sense?

Reply:
I agree on the decoupling of code in the reconciler and the validator here. But maybe it's better to implement it in the MWH and drop these VWH rules entirely?

Reply:
That sounds reasonable too at this point if it's all going to be temporary.

Reply:
I'm kinda hesitant on whether we should use VWH or MWH.

Please CMIIW - seems if we compare the ns labels in the VWH, we have to call the API server once to get the namespace object. MWH saves us 1 api call since it updates the ns object in the request directly. If it's the case, seem MWH is the better option.

WDYT @rjbez17  ?
(anyway if we decide to replace it with MWH, I will make the changes in the next PR.)

Reply:
I'm good with that, thanks.